blog post about relay changes (#234)
bluesky-social/bsky-docs: blog/2024-11-18-relay-ops.md

---
slug: relay-ops
title: Relay Operational Updates
tags: ['firehose']
---

> **Summary:** we are making a couple changes to the Bluesky Relay servers **this week**. The wire protocol and semantics are not impacted, but the changes may impact firehose consumers: 

Reviewer: ```suggestion
**Summary:** We are making a couple of changes to the Bluesky Relay servers **this week**. The wire protocol and semantics are not impacted, but the changes may impact firehose consumers: 
```

> - one way or another, event stream sequence will change for all subscribers. Most consumers can probably just reset to the new cursor; see details below.

Reviewer: ```suggestion
- One way or another, the event stream sequence will change for all subscribers. Most consumers can probably just reset to the new cursor; see details below.
```

> - the `bsky.network` hostname will be swapped to a new relay instance (with new sequence)

Reviewer: ```suggestion
- The `bsky.network` hostname will be swapped to a new relay instance (with new sequence)
```

> The atproto networking is growing rapidly! This is exciting and positive, but means we need to move up some of our network scaling plans. One particular resource bottleneck is the Relay firehose. We are seeing sustained traffic of over 2,000 events per second, and have hundreds of active consumers. Multiplied, the overall throughput is hard for any one server to keep up with.

Reviewer: ```suggestion
The atproto network is growing rapidly! This is exciting and positive, but it means we need to move up some of our network scaling plans. One particular resource bottleneck is the Relay firehose. We are seeing sustained traffic of over 2,000 events per second, and we have hundreds of active consumers. Multiplied, the overall throughput is hard for any one server to keep up with.
```

> We saw this problem coming from the beginning, and have several plans and options to mitigate it. Details for each of these are linked or in sections below.

Reviewer: ```suggestion
We saw this problem coming from the beginning and we have several plans and options to mitigate it. Details for each of these are linked or in sections below.
```

First, we are simply upgrading the size of server that we run our primary Relay instances on. We expect to do this in coming days. The main impact here will be a reset of the firehose sequence (including cursor values), which will impact all downstream consumers.

Second, we are introducing the concept of "firehose fan-out" services, and releasing our implementation [Rainbow](https://github.com/bluesky-social/indigo/tree/main/cmd/rainbow). These are servers which re-broadcast the event stream firehose to many clients, reducing the bandwidth load on relays themselves. Rainbow is now in production, as an implementation detail.

Third, we recently released [Jetstream](https://docs.bsky.app/blog/jetstream) as a more informal and light-weight option for event stream consumption. We encourage developers to check out Jetstream, and switch over if it works for their use case.

We encourage folks to take advantage of Rainbow and Jetstream. You can run your own, or make use of our hosted instances. Some additional longer-term options are discussed at the bottom of this article.

> We don't have hard time or date estimate for the cursor sequence change, but expect it this week, possibly tomorrow (Tuesday). We generally try to give more time and make changes less disruptive, but are not able to do so in this situation. Please follow the [`@atproto.com`](https://bsky.app/profile/atproto.com) account, this blog, and in [Github Discussions](https://github.com/bluesky-social/atproto/discussions) for announcements.

Reviewer: ```suggestion
We don't have a hard time or date estimate for the cursor sequence change, but we expect it this week, possibly tomorrow (Tuesday). We generally try to give more time and make changes less disruptive but we're not able to do so in this situation. Please follow the [`@atproto.com`](https://bsky.app/profile/atproto.com) account, this blog, and in [Github Discussions](https://github.com/bluesky-social/atproto/discussions) for announcements.
```

## Relay Upgrade

The current Bluesky relay instance is available at `wss://bsky.network`. A more specific hostname for this relay instance is (XXX: public old relay hostname).

> Instead of upgrading the relay in-place, we are going to start a new relay from scratch, which means event stream ordering will be different and sequence numbers will not align with the current relay. This new instance will run at (XXX: public new relay hostname).

Reviewer: ```suggestion
Instead of upgrading the relay in-place, we are going to start a new relay from scratch. This means that event stream ordering will be different and sequence numbers will not align with the current relay. This new instance will run at (XXX: public new relay hostname).
```

There will be at least a short window when both relays are running, so downstream consumers can update their services to "cut over". At some point soon after the new relay is deployed, we will switch the generic `bsky.network` hostname to point at it.

What firehose consumers need to do about the cut over depends on how careful they need to be about processing every single event on the firehose.

Services which always connect to the firehose from the current offset don’t need to do anything. This includes developer tools like `goat`, and real-time sampling like Firesky.

> Best-effort consumers can likely be configure to consume from the current firehose offset, and simply restart, missing at most a few seconds of events. This probably makes sense for feed generators, analytics, automated labeling, etc.

Reviewer: ```suggestion
Best-effort consumers can likely be configured to consume from the current firehose offset, and simply restart, missing at most a few seconds of events. This probably makes sense for feed generators, analytics, automated labeling, etc.
```

> Services which want to really minimize the number of missed events can take a different approach. They can be programmed to connect to the new relay with no cursor, detect the current sequence number, then reconnect with a cursor a couple minutes back (at current rates this means a few hundred thousand events back). This results in a time overlap window between the two relays (with some events processed twice), which helps minimize the chance of missing any events.

Reviewer: ```suggestion
Services which want to really minimize the number of missed events can take a different approach. They can be programmed to connect to the new relay with no cursor, detect the current sequence number, then reconnect with a cursor a couple of minutes back (at current rates this means a few hundred thousand events back). This results in a time overlap window between the two relays (with some events processed twice), which helps minimize the chance of missing any events.
```

## Firehose Fan-out and Rainbow

> Firehose fan-out servers have a single upstream connection to a relay firehose, and re-broadcast that stream to multiple subscribing clients. They maintain a local backfill window (allowing re-connection with a cursor), but do not implement endpoints like `getRecord` or `getRepo`, only `com.atproto.sync.subscribeRepos`. They do not re-validate the event stream (eg, don't verify signatures). They maintain the exact sequence numbers of the upstream relay, meaning that they are interchangeable with the relay and with sibling fan-out servers, and can be placed behind a load-balancer.

Reviewer: ```suggestion
Firehose fan-out servers have a single upstream connection to a relay firehose, and re-broadcast that stream to multiple subscribing clients. They maintain a local backfill window (allowing re-connection with a cursor), but do not implement endpoints like `getRecord` or `getRepo`, only `com.atproto.sync.subscribeRepos`. They do not re-validate the event stream (eg, they don't verify signatures). They maintain the exact sequence numbers of the upstream relay, meaning that they are interchangeable with the relay and with sibling fan-out servers, and can be placed behind a load-balancer.
```

Today we are releasing Rainbow, a fan-out service implemented in Go. The [source code is available](https://github.com/bluesky-social/indigo/tree/main/cmd/rainbow) in the `indigo` git repository, with the same MIT/Apache licensing as our other open source projects.

> In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the networking over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.

Reviewer: ```suggestion
In the future, we may run additional Rainbow instances outside our primary data centers, similar to how we offer Jetstream instances today. This distributes the network over more routers, and could improve connection reliability and throughput for subscribers in regions outside North America.
```
`networking` or `network`, not sure which is correct here?

Author: I think "networking" is clearer? but neither are great

Going forward, we recommend that software which consumes from firehoses (including atproto SDKs) support HTTP redirects for WebSocket connections. This enables "pooling" behavior where a single hostname could route clients to multiple distinct servers, without use of a load-balancer.

## What Else?

As a reminder, the Relay synchronization API is the same as the PDS synchronization API, including the firehose. Relays are essential to the functionality of the atproto network: they are an operational and developer convenience, allowing consuming services to skip keeping track of which PDS instances are active.

> A classic way to scale services like the firehose is *sharding*, where the stream is split in to multiple parallel streams. atproto has a natural sharding key (the account DID), which means related events can be consistently routed to the correct shard. This will allow near-indefinite scaling of the firehose event rate, assuming compute and network resources are available.

Reviewer: ```suggestion
A classic way to scale services like the firehose is *sharding*, where the stream is split into multiple parallel streams. atproto has a natural sharding key (the account DID), which means related events can be consistently routed to the correct shard. This will allow near-indefinite scaling of the firehose event rate, assuming compute and network resources are available.
```
`near-indefinite scaling` or `near-infinite scaling`, I'm not sure which would be clearer here? 🤔

> Part of what makes Relays difficult to operate at scale is that they function as both a "relay" (rebroadcasting events from PDS instances) and a full-network mirror (storing all repo contents). In the current protocol, it is necessary to combine both functions to fully verify repository operations, especially deletion of events. We think that clever implementations could make this work less resource intensive (for example, but not storing records themselves, just CIDs). It is also possible to validate *most* aspects of the event stream without a full copy of the repo tree, and there could be a role in the network for "non-mirroring" relays.

Reviewer: `(for example, but not storing records themselves, just CIDs)`
I think there might be a word or more missing here?

Author: fixed
